casper/src/main/resources/PoS.rho

```diff
@@ -1,180 +1,165 @@
+                  } else {
+                    deposit!(userPk, amount, *depositCh) |
+                    for(@depositResult <- depositCh) {
+                      bondsCh!(
+                        bonds.set(
+                          userPk,
+                          [amount, bonds.size() + 1]
+                        )
+                      ) |
+                      returnCh!(depositResult)
                     }
                   }
                 }
               }
+            } |
 
-              //Assuming that user public key is valid, a vault should always be returned.
-              | contract deposit (@amount, return) = {
-                new parametersCh,
-                  vaultCh,
-                  revAddressCh,
-                  authKeyCh in {
-                  getParameters!(*parametersCh)
-                  | for (_, _, @userPk, _ <- parametersCh) {
-                    revAddressOps!("fromPublicKey", userPk, *revAddressCh)
-                    | for (@revAddress <- revAddressCh) {
-                      @RevVault!("findOrCreate", revAddress, *vaultCh)
-                      | for (@(true, vault) <- vaultCh) {
-                        @RevVault!("deployerAuthKey", *authKeyCh)
-                        | for (userAuthKey <- authKeyCh){
-                          @vault!("transfer", posRevAddress, amount, *userAuthKey, *return)
-                        }
-                      }
-                    }
-                  }
-                }
-              }
+            contract PoS (@"pay", @amount, returnCh) = {
+              new vaultCh, transferCh, userCh in {
+                getUser!(*userCh) |
 
-              | paymentStore!([Nil, Nil, Nil, Nil])
-
-              //Make it easy to query the last payment so that validators
-              //can check how much phlo there is for the deploy.
-              | contract PoS(@"lastPayment", return) = {
-                for(@payment <- paymentStore) {
-                  paymentStore!(payment) | return!(payment)
+                for(@userPk <- userCh) {
+                  deposit!(userPk, amount, *returnCh)
                 }
               }
-
-              | contract PoS (@"pay", @amount, return) = {
-                new parametersCh, depositRet in {
-                  deposit!(amount, *depositRet)
-                  | getParameters!(*parametersCh)
-                  | for (@depositResult <- depositRet; @codeHash, @phloRate, @userPk, @timestamp <- parametersCh; lastPayment <- paymentStore) {
-                    match depositResult {
-                      (true, Nil) => { paymentStore!([codeHash, userPk, timestamp, phloRate * amount]) | return!(phloRate *amount) }
-                      (false, err) => { paymentStore!([codeHash, userPk, timestamp, 0]) | return!(0) }
-                    }
-                  }
+            }
+          }
+        } |
+
+        contract deposit (@fromUserPk, @amount, returnCh) = {
```

Comment:
Interesting; so we do a "push" transfer (as opposed to "pull" transfer) from within the PoS contract. Basically making use of the transitive dependency attack.